google/cloud/bigquery/schema.py

```diff
@@ -921,3 +915,83 @@ def from_api_repr(cls, resource: dict) -> SerDeInfo:
         config = cls("")
         config._properties = copy.deepcopy(resource)
         return config
+
+
+class Schema:
+    # TODO docstrings and type hints
+    def __init__(self, fields=None, foreign_type_info=None):
+        self._properties = {}
+        self._fields = [] if fields is None else list(fields)  # Internal List
+        self.foreign_type_info = foreign_type_info
+
+    @property
+    def foreign_type_info(self) -> Any:
+        """TODO: docstring"""
+        return self._properties.get("foreignTypeInfo")
+
+    @foreign_type_info.setter
+    def foreign_type_info(self, value: str) -> None:
+        value = _isinstance_or_raise(value, str, none_allowed=True)
+        self._properties["foreignTypeInfo"] = value
+
+    @property
+    def _fields(self) -> Any:
+        """TODO: docstring"""
+        return self._properties.get("_fields")
+
+    @_fields.setter
+    def _fields(self, value: list) -> None:
+        value = _isinstance_or_raise(value, list, none_allowed=True)
+        self._properties["_fields"] = value
+
+    def __len__(self):
+        return len(self._fields)
+
+    def __getitem__(self, index):
+        return self._fields[index]
+
+    def __setitem__(self, index, value):
+        self._fields[index] = value
+
+    def __delitem__(self, index):
+        del self._fields[index]
+
+    def __iter__(self):
+        return iter(self._fields)
+
+    def __str__(self):
+        return f"Schema({self._fields}, {self.foreign_type_info})"
+
+    def __repr__(self):
+        return f"Schema({self._fields!r}, {self.foreign_type_info!r})"
+
+    def append(self, item):
+        self._fields.append(item)
+
+    def extend(self, iterable):
+        self._fields.extend(iterable)
+
+    def to_api_repr(self) -> dict:
+        """Build an API representation of this object.
+
+        Returns:
+            Dict[str, Any]:
+                A dictionary in the format used by the BigQuery API.
+        """
+        return copy.deepcopy(self._properties)
```

Comment:
It might be a bit wasteful to make deepcopy here and in `from_api_repr`. Indeed it's safer, but could add a lot of overhead. IIRC we actually removed some `deepcopy` calls from `SchemaField` because it was slowing down customers who build dynamic schemas in their code.

See: https://github.com/googleapis/python-bigquery/issues/6 and https://github.com/googleapis/python-bigquery/pull/26